### ISSUE TITLE
"Failed to run the package service" on Synology DS116 with DSM 6.1.3-15152

### ISSUE THREAD

**Issue by ilmtr** [duplicati/duplicati#2610]
I have:
- [X] searched open and closed issues for duplicates

----------------------------------------

### Version info

**Duplicati Version:** 2.0.1.73
**Operating System:** DSM 6.1.3-15152 Update 1
**Backend:** Synology DS116

### Bug description
After installing the Duplicati package for Synology through the DiskStation Package Center Duplicati doesnt run and shows a "Failed to run the package service" error after pressing Run. 

### Steps to reproduce
As above. 

**Actual result:** "Failed to run the package service" error.
**Expected result:** Duplicati should run normally. 

### Screenshots
![untitled](https://user-images.githubusercontent.com/11330859/28568054-c2042fb2-7134-11e7-8b08-4608a4bf931a.png)

**Comment by kenkendk** [comment]
Have you installed the "Mono" package?

**Comment by ilmtr** [comment]
I found the Mono package and installed it, has status Running. 

After starting Duplicati seems to be running from the Package Center. But on opening I get a Connection lost message. Stopping / restarting didnt work. 

I then completely uninstalled Duplicati from the Package Center. And downloaded and installed the most recent version (duplicati-2.0.2.1_beta_2017-08-01.spk). Same Connection lost message...

![image](https://user-images.githubusercontent.com/11330859/28824911-7c6f9122-76c4-11e7-903a-18bb2e4dbf9f.png)

![image](https://user-images.githubusercontent.com/11330859/28824924-8d3149b0-76c4-11e7-8607-7ce1898a2398.png)

![image](https://user-images.githubusercontent.com/11330859/28824941-a252bffe-76c4-11e7-951d-138133d553c1.png)

**Comment by kenkendk** [comment]
Try reloading the page (the entire page with DSM on it).

**Comment by ilmtr** [comment]
I reloaded the page, same message. I rebooted the NAS, same message.

**Comment by tedcarnahan** [comment]
I also have this issue, same config except on a DS216j

**Comment by jake404** [comment]
Same here and I have a DS916+.

**Comment by kenkendk** [comment]
@jake404 Yes, that could be it. Are your user an administrator on the Synology box?
Duplicati requires that an Administrator is logged in, as you get full access to the file system.

**Comment by wjansenw** [comment]
I have the same issue. Worked ok before. Seems a dsm update triggered this? Maybe synology changed something wrt. permissions. Can't check now.

**Comment by jake404** [comment]
Yes, my user is an administrator. Where are the logs stored? What is the command to start the server manually from a terminal?

**Comment by wjansenw** [comment]
Duplicati is running as user root, so should have all necessary authorization I assume. Can't find an error in the Duplicati-server.sqlite (in fact, it does not show anything in the Log table) 

`user@syno:~$ ps -edf | grep mono`
`http      3658     1  0 Aug10 ?        00:00:01 /var/packages/Mono/target/usr/local/bin/mono /var/packages/Mono/target/usr/local/lib/mono/4.5/mod-mono-server4.exe --filename /tmp/mod_mono_server_global --nonstop --master`
`root     24138     1 10 20:31 ?        00:00:08 mono /volume1/@appstore/Duplicati/Duplicati.Server.exe`

**Comment by jake404** [comment]
I ran that process check and I only got this:
root     11756     1  2 20:57 ?        00:00:02 mono /volume1/@appstore/Duplicati/Duplicati.Server.exe

**Comment by wjansenw** [comment]
i stopped the duplicati package and started duplicati server manually:
`mono /volume1/@appstore/Duplicati/Duplicati.Server.exe --webservice-port=8200 --webservice-interface=any`

browsing to http://syno:8200 seems to be working fine now

**Comment by wjansenw** [comment]
so, it must be how duplicati is integrated in the DSM

**Comment by jake404** [comment]
I just tried the same thing and it works for me as well.

**Comment by kenkendk** [comment]
@wjansenw  and @jake404 please take care with that approach as it exposes Duplicati to the network (potentially the internet).

This way of invoking the server will enable the DSM integrated login support, and shield off access by checking that the user is logged into DSM:

```
export SYNO_LOGIN_CGI=/usr/syno/synoman/webman/login.cgi
export SYNO_AUTHENTICATE_CGI=/usr/syno/synoman/webman/modules/authenticate.cgi
export SYNO_DSM_AUTH=1
mono /volume1/@appstore/Duplicati/Duplicati.Server.exe --webservice-port=8200 --webservice-interface=any
```

This will still make the service available from the outside, but now require DSM login.

There are some extra variables that can further configure the setup:
https://github.com/duplicati/duplicati/blob/master/Installer/Synology/scripts/start-stop-status

You can also see how the integration with the nginx server is done.

**Comment by wjansenw** [comment]
Kenneth,
Thanks for the feedback. However, running it like that gives the same error
as within DSM.
Regards,
Wim

Op za 12 aug. 2017 om 15:34 schreef Kenneth Skovhede <
notifications@github.com>:

**Comment by DcR-NL** [comment]
**Duplicati Version:** 2.0.2.1
**Operating System:** DSM 6.1.3-15152 Update 3
**Backend:** Synology DS713+

Looks like the same issue here on newer version, also using a user in the administrators group.

After installing seeing "Failed to run the package service" error when trying to run. Installing Mono package by Synology fixed this. Now it runs, but when opening view I see a "Connection lost" error with the message: "The connection to the server is lost, attempting again in 0:12 ...". Underneath there is another dialog but I cant see the message.

**Comment by octplane** [comment]
Same here, I have checked the process run as root:

```shell
$ ps fauxw | grep -i 'du[p]li'
root     26728  0.7  5.9 1132516 55124 ?       Sl   09:02   0:02 mono /volume1/@appstore/Duplicati/Duplicati.Server.exe
```

Env variables of the process:
```
SYNO_ALL_USERS=0
SYNO_LOGIN_CGI=/usr/syno/synoman/webman/login.cgi
SYNO_AUTHENTICATE_CGI=/usr/syno/synoman/webman/modules/authenticate.cgi
SYNO_DSM_AUTH=1
TEMP=/volume1/@appstore/Duplicati/temp
TMP_DIR=/volume1/@appstore/Duplicati/temp
SYNO_SKIP_XSRF=0
SYNOPKG_DSM_LANGUAGE=enu
TERM=vt102
SYNOPKG_TEMP_LOGFILE=/tmp/synopkgmgr.log-Duplicati
SYNOPKG_USERNAME=admin
SYNOPKG_PKG_STATUS=START
SYNOPKG_DSM_ARCH=braswell
PATH=/sbin:/bin:/usr/sbin:/usr/bin:/usr/syno/sbin:/usr/syno/bin:/usr/local/sbin:/usr/local/bin
SELF=Duplicati
PWD=/
SYNOPKG_PKGDEST=/volume1/@appstore/Duplicati
SYNOPKG_PKGVER=2.0.2.1
SYNOPKG_DSM_VERSION_MINOR=1
HOME=/root
SHLVL=2
SYNOPKG_DSM_VERSION_BUILD=15152
UPSTART_INSTANCE=
SYNOPKG_PKGNAME=Duplicati
SYNOPKG_DSM_VERSION_MAJOR=6
SYNOPKG_PKGDEST_VOL=/volume1
UPSTART_JOB=pkgctl-Duplicati
_=/usr/local/bin/mono
```

On the web frontend, behind the `connection lost` dialog, there is the dialog asking to run as administrator.

Anything we can do to help debug?

**Comment by thomas-frantz** [comment]
Based on it running in root, and the fact that the API was returning 403s, I was able to fix my `connection lost` errors by adding my administrator user to the root group via SSH. Now it seems to work great.

I'm not sure how ideal of a fix this is (security or practicality wise), but my NAS is also locked down pretty heavily (The only thing public facing is the HTTPS interface, which is restricted to administrators only, and all admins require 2FA.)

**Comment by tedcarnahan** [comment]
I can confirm @thomas-frantz workaround is effective.

**Comment by tedcarnahan** [comment]
Okay, I take back my last comment - the workaround made it possible for me to connect to the server with the UI, but it still wasn't working.  In setting up a new connection, fields like encryption methods and backends were not being populated.

**Comment by thomas-frantz** [comment]
@tedcarnahan it's possible the workaround fixed that one specific issue, but your install has multiple issues. Mine is populating everything, though I haven't had a chance to setup a backup schedule yet (I have done a test connection to Backblaze B2, though.)

**Comment by tedcarnahan** [comment]
@thomas-frantz I think you're right.  I went through the system and tried to get things back to a standard configuration and duplicati is working great.  So your workaround is, in fact, effective.

Does Duplicati run as root just because of the need to access all files for the backup, or are there other reasons besides?

**Comment by wjansenw** [comment]
How do I add my user to the root group? The `groupadd` command is not available?

@tedcarnahan : I had the same issue (fields not populating). What did you do to fix that?

**Comment by davecorrie** [comment]
I had this issue and solved it by adding the 'admin' user (the account that I use to log into synology DSM) to the 'http' user group.

I did this by logging in as admin, opening DSM Control Panel -> User -> 'admin' -> Edit -> User groups. My admin user was a member of 'administrators' and 'users' groups but not 'http'. So I clicked add to group 'http' and clicked ok and Duplicati seems to work fine.

**Comment by wjansenw** [comment]
Tnx, solved it for me!!

**Comment by ilmtr** [comment]
Adding admin to the http user group also solved this problem. I was getting an SSL eror on adding an external location (Webdav / B2). Local backup worked though. This error looks similar to this one: #2711

**Comment by kenkendk** [comment]
Strangely, my user is not part of the "http" group. I checked the `/etc/group` and my user is only part of "administrator" and "user" groups, which is what I see in the DSM user manager.

**Comment by StanBoyet** [comment]
Still having the `Failed to run the package service`, even after adding the admin to the `http` user group.
![screenshot 2017-09-25 08 42 09](https://user-images.githubusercontent.com/1906401/30795348-6fde92d0-a1cd-11e7-91b0-1d7e202bcc3d.png)

**Comment by lseg74** [comment]
Hi Guys,

I tried the http group, but that did not work for me. What I actually noticed could maybe help: for me it does not work when I use the quickconnect.to . So when I remotely access my NAS (using quickconnect.to) and run duplicati in the console, ot gives the errors as described. When I access the NAS on a local machine, using the IP address instead of quickconnect, it does work (first had an XSRV (?) error, but that disapeared when i did a reload). I did not try a backup, still have to setup an SFTP server on my linux machine.

**Comment by ilmtr** [comment]
I just updated DSM to the latest version. It seems to work now without (almost) any issues. I was able to setup a backup (local target), run it and make a restore. 

Duplicati Version: duplicati-2.0.2.1_beta_2017-08-01.spk
Operating System: DSM 6.1.3-15152 Update 8
Backend: Synology DS116

The only things are during the backup process I get this message:
![ds116](https://user-images.githubusercontent.com/11330859/31774488-37257e64-b4e6-11e7-871d-e316ce18c85b.png)
Which seems to go away after completing the backup. 

Also I get an 'Error in updating':
`System.Net.WebException: Error: NameResolutionFailure
  at System.Net.WebClient.DownloadFile (System.Uri address, System.String fileName) [0x000c9] in <5071a6e4a4564e19a2eda0f53e42f9bd>:0 
  at System.Net.WebClient.DownloadFile (System.String address, System.String fileName) [0x00019] in <5071a6e4a4564e19a2eda0f53e42f9bd>:0 
  at (wrapper remoting-invoke-with-check) System.Net.WebClient:DownloadFile (string,string)
  at Duplicati.Library.AutoUpdater.UpdaterManager.CheckForUpdate (Duplicati.Library.AutoUpdater.ReleaseType channel) [0x000ee] in <2e3a1be277654a188f84e503094300c3>:0 `

`System.Net.WebException: Error: TrustFailure (The authentication or decryption has failed.) ---> System.IO.IOException: The authentication or decryption has failed. ---> System.IO.IOException: The authentication or decryption has failed. ---> Mono.Security.Protocol.Tls.TlsException: Invalid certificate received from server. Error code: 0xffffffff800b010a
  at Mono.Security.Protocol.Tls.RecordProtocol.EndReceiveRecord (System.IAsyncResult asyncResult) [0x00040] in <1d0bb82c94e7435eb09324cf5ef20e36>:0 
  at Mono.Security.Protocol.Tls.SslClientStream.SafeEndReceiveRecord (System.IAsyncResult ar, System.Boolean ignoreEmpty) [0x00000] in <1d0bb82c94e7435eb09324cf5ef20e36>:0 
  at Mono.Security.Protocol.Tls.SslClientStream.NegotiateAsyncWorker (System.IAsyncResult result) [0x00071] in <1d0bb82c94e7435eb09324cf5ef20e36>:0 
   --- End of inner exception stack trace ---
  at Mono.Security.Protocol.Tls.SslClientStream.EndNegotiateHandshake (System.IAsyncResult result) [0x0003b] in <1d0bb82c94e7435eb09324cf5ef20e36>:0 
  at Mono.Security.Protocol.Tls.SslStreamBase.AsyncHandshakeCallback (System.IAsyncResult asyncResult) [0x0000c] in <1d0bb82c94e7435eb09324cf5ef20e36>:0 
   --- End of inner exception stack trace ---
  at Mono.Security.Protocol.Tls.SslStreamBase.EndRead (System.IAsyncResult asyncResult) [0x00057] in <1d0bb82c94e7435eb09324cf5ef20e36>:0 
  at Mono.Net.Security.Private.LegacySslStream.EndAuthenticateAsClient (System.IAsyncResult asyncResult) [0x00011] in <5071a6e4a4564e19a2eda0f53e42f9bd>:0 
  at Mono.Net.Security.Private.LegacySslStream.AuthenticateAsClient (System.String targetHost, System.Security.Cryptography.X509Certificates.X509CertificateCollection clientCertificates, System.Security.Authentication.SslProtocols enabledSslProtocols, System.Boolean checkCertificateRevocation) [0x0000e] in <5071a6e4a4564e19a2eda0f53e42f9bd>:0 
  at Mono.Net.Security.MonoTlsStream.CreateStream (System.Byte[] buffer) [0x0006a] in <5071a6e4a4564e19a2eda0f53e42f9bd>:0 
   --- End of inner exception stack trace ---
  at System.Net.WebClient.DownloadFile (System.Uri address, System.String fileName) [0x000c9] in <5071a6e4a4564e19a2eda0f53e42f9bd>:0 
  at System.Net.WebClient.DownloadFile (System.String address, System.String fileName) [0x00019] in <5071a6e4a4564e19a2eda0f53e42f9bd>:0 
  at (wrapper remoting-invoke-with-check) System.Net.WebClient:DownloadFile (string,string)
  at Duplicati.Library.AutoUpdater.UpdaterManager.CheckForUpdate (Duplicati.Library.AutoUpdater.ReleaseType channel) [0x000ee] in <2e3a1be277654a188f84e503094300c3>:0 `

**Comment by ilmtr** [comment]
A SSL Webdav location as a target still doesnt work:

> Failed to connect: The server certificate had the error RemoteCertificateChainErrors and the hash BC68D5BF3D3D52537C51D6049EA43BBBF4696B05 If you trust this certificate, use the commandline option --accept-specified-ssl-hash=BC68D5BF3D3D52537C51D6049EA43BBBF4696B05 to accept the server certificate anyway. You can also attempt to import the server certificate into your operating systems trust pool. You may want to import a set of trusted certificates into the Mono certificate store. Use the command: cert-sync /etc/ssl/certs/ca-certificates.crt #for Debian based systems cert-sync /etc/pki/tls/certs/ca-bundle.crt #for RedHat derivatives Read more: http://www.mono-project.com/docs/about-mono/releases/3.12.0/#cert-sync

**Comment by ilmtr** [comment]
Any updates?  #2824 Similar issue?

**Comment by kblomster** [comment]
I experienced the "can't connect" issue as well, with Duplicati claiming an admin account was needed, but I found that adding the user to the http group is not necessary - adding any group at all fixes the issue. I haven't debugged it, but a suspicion is that when the group numbers are retrieved in [SynologyAuthenticationHandler.cs on line 200](https://github.com/duplicati/duplicati/blob/9593edae149ef88dab315cf4a8fd65b7b6adcdd0/Duplicati/Server/WebServer/SynologyAuthenticationHandler.cs#L200), the the output of `id -G 'yourusername'` is not whitespace trimmed before getting passed to split/contains on the next line, so there might be a trailing `\n` or somesuch if group number 101 happens to be the last item.

**Comment by kblomster** [comment]
@ilmtr To resolve the SSL issues, see the wiki page [SSL/TLS support in Mono](https://github.com/duplicati/duplicati/wiki/SSL-TLS-support-in-Mono) - instructions for Synology DSM at the bottom.

**Comment by ilmtr** [comment]
I should be fixed with an updated version of Mono from SynoCommunity, not Synology (=older version?). Havent checked myself if it works 'out of the box' already...

https://github.com/SynoCommunity/spksrc/issues/3033

**Comment by FloLaco** [comment]
I've just installed Duplicati and I got the same error.
After reading the @kblomster comment  (https://github.com/duplicati/duplicati/issues/2610#issuecomment-358072917), I checked the source code and my situation.
The nginx server returned me "403 Administrator login required".

I checked on my NAS : 
```
ash-4.3# id -G 'Rian' | cat -e
100 101$
```

After adding a group on my account (from Synology interface), here's the new output : 
ash-4.3# id -G 'Rian' | cat -e
100 101 1023$

And now I don't have the issue anymore ...

@kblomster is right because `trim()` does not remove "\n", and the groups contain "101\n" not "101" (I guess, I've no dump the variable).
